incubator/hnc/internal/reconcilers/object_test.go

```diff
@@ -231,6 +231,118 @@ var _ = Describe("Exceptions", func() {
+			name:      "not propagate object to a newly-unlabeled child",
+			toUnlabel: labeledchild,
+			want:      []string{},
+			notWant:   []string{labeledchild, nolabelchild},
+		}, {
+			name:       "propagate object to a new child with the label",
+			toAddChild: labeledns,
+			want:       []string{labeledchild, labeledns},
+			notWant:    []string{nolabelchild},
+		}, {
+			name:       "not propagate object to a new child without the label",
+			toAddChild: nolabelns,
+			want:       []string{labeledchild},
+			notWant:    []string{nolabelchild, nolabelns},
+		}}
+
+		for _, tc := range tests {
+			// Making a local copy of tc is necessary to ensure the correct value is passed to the closure,
+			// for more details look at https://onsi.github.io/ginkgo/ and search for 'closure'
+			tc := tc
+			It("Should "+tc.name, func() {
+				// Set up namespaces
+				names := map[string]string{
+					p:            createNS(ctx, p),
+					labeledchild: createNSWithLabel(ctx, labeledchild, map[string]string{label: "true"}),
+					nolabelchild: createNS(ctx, nolabelchild),
+					labeledns:    createNSWithLabel(ctx, labeledns, map[string]string{label: "true"}),
+					nolabelns:    createNS(ctx, nolabelns),
+				}
+				setParent(ctx, names[labeledchild], names[p])
+				setParent(ctx, names[nolabelchild], names[p])
+
+				// Create a Role and verify it's propagated to all children.
+				makeObject(ctx, api.RoleResource, names[p], "testrole")
+				Eventually(hasObject(ctx, api.RoleResource, names[labeledchild], "testrole")).Should(BeTrue(), "When propagating testrole to %s", names[labeledchild])
+				Eventually(hasObject(ctx, api.RoleResource, names[nolabelchild], "testrole")).Should(BeTrue(), "When propagating testrole to %s", names[nolabelchild])
+				// Add `select` exception annotation with propagate label and verify the
+				// object is only propagated to children with the label.
+				updateObjectWithAnnotation(ctx, api.RoleResource, names[p], "testrole", map[string]string{
+					api.AnnotationSelector: label,
+				})
+				Eventually(hasObject(ctx, api.RoleResource, names[nolabelchild], "testrole")).Should(BeFalse(), "When propagating testrole to %s", names[nolabelchild])
+				Consistently(hasObject(ctx, api.RoleResource, names[labeledchild], "testrole")).Should(BeTrue(), "When propagating testrole to %s", names[labeledchild])
+
+				// ?????????????
+				// For issue 1448, we may need to add a sleep to ensure the source object
+				// reconciliation is done and not enqueued elsewhere, but even with 360s
+				// sleep, the test still passed before the fix for 1448 (manual tests before
+				// the fix failed).
+				//time.Sleep(60 * time.Second)
+
+				// Add the label to the namespace if the value is not empty.
+				if tc.toLabel != "" {
+					ns := getNamespace(ctx, names[tc.toLabel])
+					ns.SetLabels(map[string]string{label: "true"})
+					updateNamespace(ctx, ns)
+				}
```

Comment:
Ah figured out the issue! It's the problem in setting labels. Here it resets the labels, including the tree labels! That's why the parent namespace still gets all object synced (because the ns is considered as `updated` even before the fix).

I will update the logic for label and unlabel here and resubmit.